---
title: BMEG 424 Assignment 3
output:
  github_document:
    toc: true
    toc_depth: 4
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# BMEG 424 Assignment 3: ChIP-seq
## Introduction:
### General Information About Assignments
We would like to begin by reminding you of some of the key rules and regulations you will have to follow for the assignments in this course:


- All assignments are to be completed individually or with one partner
  
- If you choose to work in pairs for an assignment, **ONE** of you should make a submission on canvas. In the submission you **MUST** report the name & student number of BOTH partners. 

- Each time you work with a new partner you will need to create a new GitHub repo for that assignment. You should share your repos with the @BMEGGenoInfo account on Github so TAs have access to your submission.

- Late assignments will have 10% deducted for each day they are late and assignments will not be accepted >3 days late. Making changes to the files on your repo after the assignment has been submitted on Canvas will result in a late penalty unless an instructor has specifically asked you to change something.

- Each student will have to ensure their assignment files are knit correctly (see below for instructions) and legible on GitHub. 20% or more of the assignment marks can be removed at the TA's discretion for incorrect knitting or formatting of assignment files.

- We are aware that some of you will choose to use large language models like ChatGPT to complete some portions of the assignment. You are free to use ChatGPT or other LLMs to help you on the assignment, *provided you cite the model and a link to the conversation* (see [here](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq) for how to do this with ChatGPT)

- More generally, it is a requirement that you include any and all sources from where you received help (excluding your previously mentioned partner) or information for completing any question on the assignments. If you are unsure of when/how to cite your sources (including LLMs), please get in contact with the instructors.
  
If any of these rules are unclear please get in contact with the instructors.

### Goals and Objectives
The aim of this assignment is to familiarize you with the processing and analysis of ChIP-seq data. You worked on processing and aligning ChIP-seq data in the previous assignment but you might be a little confused about what to do after the alignment.
This assignment's main aim is to walk you through a ChIP-seq pipeline **post-alignment**. You will be analyzing 3 different histone modification marks (H3K27me3, H3K4me3 and H3K27ac). In order to identify the enrichments for each epigenetic mark, we also need to use the *input* which represents the DNA content of the sheared chromatin sample prior to immunoprecipitation. 

### Data
All the files can be found under the following path: **/projects/bmeg/A3/** . The files belong to one of four categories:
  
  - H3K27me3  
  
  - H3K4me3 
  
  - H3K27ac  
  
  - input (input_chr3_subset.bam)

Unfortunately for you, as you can see from the file names (other than input_chr3_subset.bam) the sequencing center has mixed up your samples, they are labelled Sample_2, Sample_4 and Sample_6. As part of this assignment you will work on identifying which sample corresponds to which epigenetic mark is which by analyzing the files.

### Software and tools
We will be using a couple of new tools this time. Before we move on to the practical part, make sure you have them all installed. 

  - Integrative Genomics Viewer (IGV): Interactive tool to visualize different data types of genetic information (e.g. bam, bed files). After assignment 1 you should have this installed on your **local computer**. To visualize where the reads of our ChIP analysis mapped in the genome. To install it, follow the instructions on this website: *https://software.broadinstitute.org/software/igv/home*
  
  - Deeptools (https://deeptools.readthedocs.io/en/develop/index.html): Software to analyze high-throughput data that allows to create easy to visualize figures. This will be installed on the server as part of your conda environment.  
  
  - macs2 (https://github.com/macs3-project/MACS/blob/master/docs/callpeak.md): Tool to capture enrichments of ChIP-seq analysis. This will be installed on the server as part of your conda environment.

### Other notes:

- When you begin your assignment do not copy the data from `/projects/bmeg/A3/` to your home directory. You can use the files in the projects folder without modifying them. Remember to output any files you create in your home directory *and not in the projects directory*. You should gzip files while you are working on the assignment and remember to delete files you no long need. If you take up too much space in your home directory you will not be able to save your work and will prevent others from doing the same, you will also make the TAs very sad.
  
### Submission:
Submit your assignment as a knitted RMarkdown document. *Remember to specify the output as github_document* You will push your knitted RMarkdown document to your github repository (one for each group). Double check that all files (including figures) necessary for your document to render properly are uploaded to your repository. 
You will then submit the link, along with the names and student numbers of all students who worked on the assignment to the assignment 3 page on Canvas. Your assignment should be submtited, and your last commit should be made, before 11:59pm on the day of the deadline. Late assignments will will be deducted 10% per day late. Assignments will not be accepted after 3 days past the deadline.

## Experiment and Analysis:
### 1. ChIP-seq 
ChIP-seq experiments work by pulling down DNA fragments which are bound by a particular protein. The DNA fragments are then sequenced and aligned to the genome. The number of reads in a given position is proportional to the amount of DNA that was pulled down with the protein of interest (ex. H3K27me3). However, the number of reads is also dependent on the amount of DNA that was sequenced at that locus. This means that if you have a sample with a lot of DNA, you will have more reads than a sample with less DNA. To account for this, we normalize the ChIP signal by the input signal. This gives us a normalized ChIP signal that we can use to perform analyses.

### 2. Exploratory analysis with IGV (8 pts)
These histone modifications mark states of active (H3K27ac, H3K4me3) or inactive (H3K27me3) gene transcription and have different coverage of the genomic region where they are located. To better visualize the differences, we will create bigWig files from previously aligned, filtered and indexed bam files. BigWig files are indexed, compressed files that can be used to **visualize** signals across the genome. Here, we will be using them to graph the coverage across the genome. We will be creating our bigWig files using the bamCoverage tool from the [deepTools suite](https://deeptools.readthedocs.io/en/develop/content/tools/bamCoverage.html).


```{bash, eval=FALSE}
Below is the command for converting the input DNA data into a bigWig file using the bamCoverage tool from the deepTools suite. 
"""
bamCoverage \
-b input_chr3_subset.bam -o \
input_chr3_subset.bw -p 2
"""

#?# 1. Create a generalized snakemake rule for converting any bam file into a bigWig file using the code above (0.25pts) and explain why we are including the -p 2 flag (0.25pts).

rule bam_to_bigwig:
    input:
        bam="{sample}.bam"
    output:
        bigwig="{sample}.bw"
    params:
        threads=2  # Using 2 threads for efficiency
    shell:
        "bamCoverage -b {input.bam} -o {output.bigwig} -p {params.threads}"

# Explanation of -p 2 flag:
# The `-p 2` flag specifies that `bamCoverage` should use 2 threads. This helps speed up
# the computation by parallelizing the processing of BAM file conversion, making it more efficient.

```

Process all of the bam files in the directory `/projects/bmeg/A3/bam/` into bigWig files. **Remember to make the bigWig files in your home directory and NOT in the projects directory**. Whether you choose to do this manually (by directly running the command four times) or with your one-step "pipeline" is up to you. Once you're done download the bigWig files and load them into IGV. Load all the bigwig signal track files (*Tip: you can use the "File" tab to "load from file" option to choose the files from your computer at once*) onto IGV on your local computer, select the "autoscale" option for each file on their individual tracks. Use the hg38 reference genome (selectable within IGV) to visualize the bigWig files.

Navigate to the following region: `chr3:93,432,899-93,508,213` and take a screenshot of your IGV session. 
```
#?# 2. Include your screenshot below using Rmarkdown syntax (0.5pts )
```
![IGV_A3IGV_A3](/A3images/IGV_A3.png)

```
#?# 3. What do you see? Is there anything peculiar about the signals at this locus? Explain what you think is causing the effect you see and how you could confirm your hypothesis (2pts) 
# HINT: Sometimes track signal will be truncated to a pre-set maximum. If you right-click the track label (left) and select "autoscale", it will automatically adust the scales of the track to the range of your data. Try playing around with the track settings to see their effect.

The IGV screenshot displays a genomic region on chromosome 3 (chr3) between positions 93,432,899 and 93,508,213, showing distinct peaks and troughs in signal intensity. There is a sharp upward spike in the input control that in the histone samples appears as downward broader spikes, which are a peculiar distinction relative to the input. This could be due to mappability issues, such as regions with low sequence complexity or repetitive elements causing misalignment in reads. The input control may show an atrificial enrichment which is the upward wpike, while normalized ChiP-seq data may show a relative depletion due to over-correction, which is the downward spikes. If there are genomic duplications, the input control can show a higher read count due to thiese, while Chip-seq normalization reduces this. If this region is a promoter or enhancer it might be naturally depleted of nuceleosomes to alloow transcription factor binding. This can also show as the high signal in input, and low signal in ChiP. 

To confirm the hypothesis, you could compare the signals with a control sample to check for consistency, validate the findings using another method like PCR or a different sequencing platform, and review the data for potential artifacts such as PCR duplicates or mapping errors. This systematic approach would help determine the true nature of the observed signals at this locus.
```

For your convenience we normalized bigWig track files for each of the histone modifications using the input. These files are located in the directory `/projects/bmeg/A3/bigWig_norm`. We are going to use these normalized bigWig files for most of our further analysis.

While exploring the bigwig files of the epigenetic marks on IGV, you probably noticed that they can look very different from each other and some of them resemble the input more closely than others. Different epigenetic marks can have very different signatures based on their distribution patterns across the genome.

ChIP-seq reads cluster together in high density regions called peaks. These peaks peaks can look different depending on what sort of protein you are pulling down in your experiment. Certain well-characterized epigentic marks are broadly classified as either active or repressive marks and we can differentiate them by visualizing their peaks. 

The deepTools **computeMatrix scale-regions** command calculates scores to represent the reads mapping to specified regions of the genome across different files. We will use these matrices to compute heatmaps which will allow us to visualize the peaks of each epigenetic mark in our experiment at certain genes of interest.

Specifically, we will use the genes located in the `reference_genes.bed` file which is located `project/bmeg/A3` directory. This file contains the coordinates of genes for which we want to understand 

```{bash, eval=FALSE}
#?# 4. Use computeMatrix to compute a matrix for the signal tracks for each histone modification outlined above (which we will use to create a plot in the following step), with the following criteria: 

## - Use the regions in reference_genes.bed located under the /projects/bmeg/A3/ directory as the basis for the plot.
## - Include the surrounding regions; 1kb upstream and 1kb downstream
## - Use all 3 input-normalized bigWig files as signal tracks (i.e. one command for all 3 marks)
## - Ensure all regions in bigWig files are are size-equalized/scaled when computing the matrix 
Write the command you used to run it below (1 pts)
```

```
computeMatrix scale-regions \
    -S /projects/bmeg/A3/bigWig_norm/Sample_2_norm.bw \
       /projects/bmeg/A3/bigWig_norm/Sample_4_norm.bw \
       /projects/bmeg/A3/bigWig_norm/Sample_6_norm.bw \
    -R /projects/bmeg/A3/reference_genes.bed \
    --beforeRegionStartLength 1000 \
    --afterRegionStartLength 1000 \
    --regionBodyLength 2000 \
    --skipZeros \
    -o ~/conda_envs/A3_env/pipeline/matrix.gz \
    --outFileSortedRegions ~/conda_envs/A3_env/pipeline/sorted_regions.bed
```

Now that the scores matrix has been computed we can use it to create a heatmap to visualize the distribution of reads at our reference genes. We will use the deepTools **plotHeatmap** function to create a heatmap following this criteria: 
```{bash, eval=FALSE}
#?# 5. Use the deepTools ** plotHeatmap ** function to create a heatmap following this criteria (1 pts) 
- Use the matrix from the previous point
- Use the Blues colourmap
- Create 3 clusters within the heatmap according to the patterns of the reads distribution across the files using hierarchical clustering
```

```
plotHeatmap -m ~/conda_envs/A3_env/pipeline/bed/matrix.gz \
    -out ~/conda_envs/A3_env/pipeline/heatmap/histone_marks_heatmap.png \
    --colorMap Blues \
    --kmeans 3
```

#?# 6. Explain what you are looking at (Axes, colours, curves) (1 pts)
```
The generated heatmap is a representation of the enrichment of histone modifications across gene regions. The x-axis represents the gene distance in bp and displays how histone modifications are distributed around genes. It goes from -1 kb upstream to the Transcription End Site (TSS) +1 kb. The y-axis depicts the genes and clusters where each row corresponds to a different gene region. The heatmap is divided into three clusters based on enrichment patterns that were determined from the k-means clustering. The darker blue color represents higher histone mark enrichment and thus areas where more ChIP-seq reads were mapped to that region, while lighter blue represents lower enrichment and fewer mapped reads in that region. The curves at the top show the average enrichment for each histone mark and show the overall trend in signal intensity across gene regions.

reference: https://academic.oup.com/nar/article/42/W1/W187/2435511
```

#?# 7. Is this plot helpful for distinguishing between the different epigenetic marks? Why or why not? (1.5 pts)
```
Yes, the plot should be useful in distinguishing epigenetic marks since the histone modifications should have different enrichment patterns. H3K4me3 should peak at TSS since it is associated with active promoters. H3K27ac may have broader enrichment upstream and downstream of genes, marking active enhancers. H3K27me3 may have a more diffused signal since it is often spreading over gene bodies and associated with repressed genes.
```

#?# 8. Include the plot produced by plotHeatmap below using Rmarkdown syntax (0.5 pts)

**Add screenshot here:**
![histone_marks_heatmap](/A3images/histone_marks_heatmap.png)

#?# 9. The above heatmap was made with the ratio of ChIP to input. Repeat the process above, but this time using the raw bigwig files (not input-normalized). 
# Include a screenshot of this heatmap, below this code block. (0.25 pt)
# How does this compare to the input-normalized data? Why do you think this is? (1pt)
```
The input normalized heatmap has clearer and more consistent patterns with distinctly enriched peaks around transcription start sites, reducing background noise and showing potential biologically relevant histone modification signals. The raw heatmap, on the other hand, has far fewer defined peaks, and more background noise, leading to fewer overall reads being mapped.
```
**Add screenshot here:**
![histone_marks_raw_heatmap](/A3images/histone_marks_raw_heatmap.png)


### 3. Peak calling (5.5 pts)
#### a. Peak calling with macs2
Now we want to identify enriched regions of the genome for each of our three histone marks. Instead of visualizing the simple distribution of reads we want to quantitatively define enriched regions or "peaks". In order to get the enrichments, we will run the **macs2** program to call the peaks for each epigenetic mark.
```{bash, eval=FALSE}
#?# 10. Type a *generalized* command (note: not a snakemake rule) for calling peaks on your bam files using macs2, using the input file as a control (0.25 pts)
## Tip: Make sure to read the documentation (using the -h flag) for the *masc2 callpeak* command

```
```
macs2 callpeak -t {chip_bam} -c {input_bam} \
    -f BAM \
    -g hs \
    -n {sample_name} \
    --outdir ~/environments/A3/pipeline/peaks \
    --nomodel --extsize 200 \
    --keep-dup all \
    -q 0.05
```
**This processing has already been completed for you.** The appropriate peak files are at `/project/bmeg/A3/*.peak`. You do not need to run the generalized command you wrote above.

#### b. Peak enrichments
For this assignment, we are working with 3 different epigenetic marks: H3K4me3, H3K27me3 and H3K27ac. We want to differentiate between these marks by comparing the relative position of the peaks called for each of them. In order to do this we will be creating a similar heatmap to the one we created above, but this time we will be visualizing the peaks against the read distributions of the other marks, instead of a set of reference genes. (i.e. H3K4me3 peaks vs H3K27me3 peaks vs H3K27ac peaks). 

```{bash, eval=FALSE}
Create 3 heatmaps following the specifications you used on part 2. In each heat map, ONE of the samples .peak files should be used as the reference file while the bigWig files of the other two samples should be used as the score files. This way you get three matrices which compare the peaks of each epigenetic mark to the reads of the other two marks. 
#?# 11. Write the commands you used to compute the matrices: (0.75 pts)
```
```
computeMatrix reference-point \
    --referencePoint center \
    -R /projects/bmeg/A3/peak/Sample_2_chr3_subset_ChIP_peaks.peak \
    -S ~/conda_envs/A3_env/pipeline/bigWig/Sample_4.bw \
       ~/conda_envs/A3_env/pipeline/bigWig/Sample_6.bw \
    --beforeRegionStartLength 1000 \
    --afterRegionStartLength 1000 \
    --binSize 50 \
    --sortRegions descend \
    -out ~/conda_envs/A3_env/pipeline/bed/sample2_peaks_matrix.gz

computeMatrix reference-point \
    --referencePoint center \
    -R /projects/bmeg/A3/peak/Sample_4_chr3_subset_ChIP_peaks.peak \
    -S ~/conda_envs/A3_env/pipeline/bigWig/Sample_2.bw \
       ~/conda_envs/A3_env/pipeline/bigWig/Sample_6.bw \
    --beforeRegionStartLength 1000 \
    --afterRegionStartLength 1000 \
    --binSize 50 \
    --sortRegions descend \
    -out ~/conda_envs/A3_env/pipeline/bed/sample4_peaks_matrix.gz

computeMatrix reference-point \
    --referencePoint center \
    -R /projects/bmeg/A3/peak/Sample_6_chr3_subset_ChIP_peaks.peak \
    -S ~/conda_envs/A3_env/pipeline/bigWig/Sample_2.bw \
       ~/conda_envs/A3_env/pipeline/bigWig/Sample_4.bw \
    --beforeRegionStartLength 1000 \
    --afterRegionStartLength 1000 \
    --binSize 50 \
    --sortRegions descend \
    -out ~/conda_envs/A3_env/pipeline/bed/sample6_peaks_matrix.gz
```
Next you'll want to create the heatmaps using the matrices you just created. You want three heatmaps (one for each matrix created in Q11). Use the same parameters you used in Q5 (Part 2). 
```{bash, eval=FALSE}
#?# 12. Write the commands you used to create the heatmaps: (0.75 pts)
```
```
plotHeatmap -m ~/conda_envs/A3_env/pipeline/bed/sample2_peaks_matrix.gz \
    -out ~/conda_envs/A3_env/pipeline/heatmap/sample2ChipSeqHeatmap \
    --colorMap Blues \
    --kmeans 3

plotHeatmap -m ~/conda_envs/A3_env/pipeline/bed/sample4_peaks_matrix.gz \
    -out ~/conda_envs/A3_env/pipeline/heatmap/sample4ChipSeqHeatmap \
    --colorMap Blues \
    --kmeans 3

plotHeatmap -m ~/conda_envs/A3_env/pipeline/bed/sample6_peaks_matrix.gz \
    -out ~/conda_envs/A3_env/pipeline/heatmap/sample6ChipSeqHeatmap \
    --colorMap Blues \
    --kmeans 3
```

#?# 13. Add screenshots of the 3 heatmaps you got using the epigenetic marks' peak files as reference files. Add them after this code chunk in the following order: sample_2, sample_4, sample_6 (the plot where sample_2 was used as the reference file is the sample_2 plot.) (0.75 pts)

**Add screenshot here:**
![sample2ChipSeqHeatmap](/A3images/sample2ChipSeqHeatmap.png)
![sample4ChipSeqHeatmap](/A3images/sample4ChipSeqHeatmap.png)
![sample6ChipSeqHeatmap](/A3images/sample6ChipSeqHeatmap.png)
#?# 14. Do you see an overlap between the peaks of different epigenetic marks? Which epigenetic marks? (1 pt)
```
Yes, there is an overlap between the peaks of sample 2 and sample 6. When sample 2 ChIP-seq peaks are used as reference, the signal form Sample 6 shows a strong peak centered at the same regions. Similarly when Sample 6 ChIP-seq peaks are used as the reference, the signal from Sample 2 also aligns. 
```

#?# 15. Why do you think these epigenetic marks overlap? (2 pt)
```
This suggests that the histone modifications corresponding to Sample 2 and Sample 6 often occur at the same genomic regions. Given that H3K4me3 and H3K27ac are both associated with active promoters and transcriptional activation it is likely these are the marks showing the co-enrichment. H3K4me3 is typically found at promoters while H3K27ac marks active enhancers and promoters.

reference: https://www.sciencedirect.com/science/article/pii/S1097276513001020#:~:text=Nucleosomes%20directly%20flanking%20TF%20binding,in%20a%20myriad%20of%20studies.
```

# Discussion (7 pts)
#?# 16. Based on the sum of your analyses which file do you believe is the H3K27me3 sample? Why? (1 pts)
```
Sample 4 is most likely to be the H3K27me3 sample due to its broad peaks covering a large genomic region. Furthermore, the lack of overlap of Sample 4 with the others reinforces this, since it is repressive it will not have overlap with H3K27ac which is active.

reference: https://pmc.ncbi.nlm.nih.gov/articles/PMC7846766/
```

#?# 17. Can you distinguish between the H3K4me3 and H3K27ac samples from the analyses you did above? (1 pt)
```
Yes, identifying Sample 6 as H3K4me3 since it is more sharply localized at TSS. In comparison H3K27ac is a bit broader since it marks both active enhancers and promoters. Though further analysis should be conducted.

reference: https://www.pnas.org/doi/full/10.1073/pnas.1016071107
```

#?# If yes, how can you tell? If not, describe (in detail) the analysis you would perform on the files to determine which is which. Remember to cite your sources (4 pts).
```
This could include Motif Enrichment Analysis as H3K27ac-marked enhancers are often enriched for AP-1 binding motifs. AP-1 transcription factors play a significant role in enhancer activation and are associated with regions marked by H3K27ac. 

reference: https://pmc.ncbi.nlm.nih.gov/articles/PMC5744881

Also the co-occupancy with RNA-seq data could be analyzed. Genes marked with H3K4me3 at their promoters are typically associated with active transcription and higher expression levels. H3K4me3 is enriched near transcription start sites and is positively correlated with gene expression.

reference: https://pmc.ncbi.nlm.nih.gov/articles/PMC3200295
```

#?# 18. For this analysis we used sheared DNA as the input/control. If we were targeting a more specific binding protein (ex. a Transcription Factor) would this still be an appropriate control? Why or why not? (1 pt)
```
Sheared DNA should not be a good control if targeting a specific binding protein like a transcription factor. This is because transcription factors find to specific DNA sequences. Using sheared DNA cannot account for background signals related to non-specific binding or chromatin accessibility.

reference: https://pubmed.ncbi.nlm.nih.gov/22955991/
```

# Contributions

Mekail(35922863): Collaborated throughout assignment.
Caden(56684616): Collaborated throughout assignment.

Please note here the team members and their contributions to this assignment.
